File: Graph/Basic/Traversals/BFS/EdgeListBFS/EdgeListBFS.cpp
// CONSIDERING TRAVERSAL FOR NON CONNECTED COMPONENTS ALSO
#include <bits/stdc++.h>
using namespace std;
void traverse(int start, vector<pair<int, int>> &edges, vector<int> &visited) // main logic
{
    queue<int> open;
    cout << start << " ";
    open.push(start);
    visited[start] = 1;
    while (!open.empty())
    {
        int curr = open.front();
        open.pop();
        for (pair<int, int> edge : edges)
        {
            if (edge.first == curr && visited[edge.second] == 0)
            {
                cout << edge.second << " ";
                visited[edge.second] = 1;
                open.push(edge.second);
            }
            if(edge.second == curr && visited[edge.first] == 0) {
                cout << edge.first << " ";
                visited[edge.first] = 1;
                open.push(edge.first);
            }
        }
    }
    cout << "\n";
}
int traversal(int V, vector<pair<int, int>> &edges)
{
    int count = 0; // number of unconnected components
    vector<int> visited(V, 0);
    for (int i = 1; i < V; i++)
    {
        if(visited[i] == 0) {
            count++;
            traverse(i, edges, visited);
        }
    }
    return count;
}
int main(void)
{
    vector<pair<int, int>> edges;
    edges.push_back(make_pair(2, 1));
    edges.push_back(make_pair(2, 4));
    edges.push_back(make_pair(4, 3));
    edges.push_back(make_pair(1, 3));
    edges.push_back(make_pair(5, 6));
    edges.push_back(make_pair(6, 7));
    edges.push_back(make_pair(5, 7));
    edges.push_back(make_pair(8, 9));
    cout << traversal(11, edges);
    return 0;
}

File: Graph/Basic/Traversals/BFS/EdgeListBFS/EdgeListBFS.java
// CONSIDERING TRAVERSAL FOR NON CONNECTED COMPONENTS ALSO
import java.util.ArrayDeque;
import java.util.Deque;

public class EdgeListBFS {
    static void traverse(int start, int edges[][], int visited[]) {
        Deque<Integer> q = new ArrayDeque<>();
        visited[start] = 1;
        q.offer(start);
        System.out.print(start+ " ");
        while(!q.isEmpty()) {
            int curr = q.poll();
            for(int[] edge: edges) {
                if(edge[0] == curr && visited[edge[1]] == 0) {
                    System.out.print(edge[1]+" ");
                    visited[edge[1]] = 1;
                    q.offer(edge[1]);
                } else if(edge[1] == curr && visited[edge[0]] == 0) {
                    System.out.print(edge[0]+" ");
                    visited[edge[0]] = 1;
                    q.offer(edge[0]);
                }
            }
        }
        System.out.println();
    }
    static void traversal(int V, int edges[][]) {
        int visited[] = new int[V+1];
        for(int i=1; i<V; i++) {
            if(visited[i] == 0) {
                traverse(i, edges, visited);
            }
        }
    }
    public static void main(String[] args) {
        int[][] edges = {{2, 1}, {2, 4}, {4, 3}, {1, 3}, {5, 6}, {6, 7}, {5, 7}, {8, 9}};
        traversal(11, edges);
    }
}
